### ISSUE TITLE
EfficientNet-Lite in SavedModel (TF2) on Hub

### ISSUE THREAD

**Issue by willbattel** [tensorflow/hub#751]
As a followup to #550, can the EfficientNet-Lite models & feature vectors be made available in TF2 format? It looks like their non-lite counterparts got ported but those of us trying to use the lite variants are out of luck when it comes to the [restrictions](https://www.tensorflow.org/hub/model_compatibility) in TF2.

**Comment by WGierke** [comment]
Hi @willbattel 
Unfortunately, I'm not aware of any such effort from the TensorFlow team or beyond, but we always welcome contributions!

That being said, if your concern is about being unable to fine-tune a Hub Module in a KerasLayer in TF2, one potential workaround you might want to try is the following:

Instead of 
```python
model_handle = "https://tfhub.dev/tensorflow/efficientnet/lite4/feature-vector/2"
model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(380, 380, 3)),
    hub.KerasLayer(model_handle, trainable=True),
    tf.keras.layers.Dense(100)  # Throws: Setting hub.KerasLayer.trainable = True is unsupported when loading from the TF1 Hub format.
])
```
one can wrap the Module in a Keras model like so:

```python
keras_layer = hub.KerasLayer(model_handle)
model = tf.keras.Sequential([keras_layer])
# Send some input through the model to generate the necessary concrete functions.
model(tf.random.uniform([1, 380, 380, 3], maxval=1))
# Save the model.
tf.saved_model.save(model, "./efficient_lite")

model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(380, 380, 3)),
    hub.KerasLayer("./efficient_lite", trainable=True),
    tf.keras.layers.Dense(100)
])
```
I've briefly confirmed that workaround with the https://www.tensorflow.org/hub/tutorials/tf2_image_retraining tutorial. Hope that helps!

**Comment by willbattel** [comment]
Thanks for the idea, I hadn't thought of that.

> Unfortunately, I'm not aware of any such effort from the TensorFlow team or beyond

I was worried that might be the case. I thought it was odd that EfficientNet-Lite was not released in TF2 format from the get-go, having been published in March 2020 [1]. Furthermore, considering how popular these lite variants are, I'm even more surprised that they *still* aren't available in TF2 format. I don't have much ground to complain on because these models are available for free, but I do think it is weird that Google hasn't made their popular state-of-the-art image classification models available on the current generation of their own platform.

Anyways, thanks for the help. Hopefully someone will eventually come along to update these models.

[1] https://blog.tensorflow.org/2020/03/higher-accuracy-on-vision-models-with-efficientnet-lite.html

**Comment by willbattel** [comment]
> I've briefly confirmed that workaround with the https://www.tensorflow.org/hub/tutorials/tf2_image_retraining tutorial.

@WGierke When you tried it, did you verify that the weights were trainable? I just tried wrapping the KerasLayer in its own model as you suggested but the number of trainable weights did not change.

```
backbone_uri = "https://tfhub.dev/tensorflow/efficientnet/lite3/feature-vector/2"
backbone_layer = hub.KerasLayer(backbone_uri, trainable=False)
model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(300, 300, 3)),
    backbone_layer,
    tf.keras.layers.Dense(443)
])
```
Total params: 7,560,251
Trainable params: 567,483
Non-trainable params: 6,992,768

```
backbone_uri = "https://tfhub.dev/tensorflow/efficientnet/lite3/feature-vector/2"
backbone_layer = hub.KerasLayer(backbone_uri)
model = tf.keras.Sequential([backbone_layer])
model(tf.random.uniform([1, 300, 300, 3], maxval=1))
tf.saved_model.save(model, "./efficient_lite")

model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(300, 300, 3)),
    hub.KerasLayer("./efficient_lite", trainable=True),
    tf.keras.layers.Dense(443)
])
```
Total params: 7,560,251
Trainable params: 567,483
Non-trainable params: 6,992,768

**Comment by MorganR** [comment]
Hello @willbattel, no unfortunately this approach does not make the weights of the underlying TF Hub module trainable in TF2. The only trainable parts are the weights for the new dense output layer you have added.

I'm keeping this issue open as a feature request for the models. Thank you for reaching out!

**Comment by alelotti96** [comment]
In the meantime, is there a way to get the EfficientNet Lite networks to be trained from scratch in TF2?

**Comment by ameyparanjape** [comment]
Interested in a similar case where I want to train EfficientNet-Lite0/1/2 models on custom data with 2 classes, any updates would be great!

**Comment by ameyparanjape** [comment]
[This](https://github.com/tensorflow/tpu/issues/905#issue-852880647)

**Comment by alelotti96** [comment]
Thanks to the authors of [this repo](https://github.com/sebastian-sz/efficientnet-lite-keras) I have been able to load EfficientNet Lite 0 with pretrained weigths!

**Comment by ameyparanjape** [comment]
@alelotti96 do you mind also sharing your experience using efficientnet lite models from this repo, training/finetuning on custom data?

**Comment by alelotti96** [comment]
@ameyparanjape I am quite new to TensorFlow, so I need some time to do everything but I will share my experience in the next few days!

**Comment by ameyparanjape** [comment]
Awesome thanks! I was able to follow the directions and get the weights converted to `.h5` format. Now I'm looking into a possibility of finetuning this on custom data. This [PR](https://github.com/tensorflow/tensorflow/pull/48400#issue-611430800) is promising.

**Comment by pindinagesh** [comment]
@willbattel 

Closing the issue as the [PR](https://github.com/tensorflow/tensorflow/pull/48400#issue-611430800) is already merged which fixes the issue. Thanks

**Comment by UsharaniPagadala** [comment]
#https://github.com/tensorflow/tensorflow/pull/48400#issue-611430800
